packages/babel-plugin-transform-typescript/src/index.ts

```diff
@@ -175,118 +179,140 @@ export default declare((api, opts) => {
+                  // injected by another plugin. Currently core does not do a good job
+                  // of keeping scope bindings synchronized with the AST. For now we
+                  // just bail if there is no binding, since chances are good that if
+                  // the import statement was injected then it wasn't a typescript type
+                  // import anyway.
+                  if (
+                    binding &&
+                    isImportTypeOnly({
+                      binding,
+                      programPath: path,
+                      pragmaImportName,
+                      pragmaFragImportName,
+                    })
+                  ) {
+                    importsToRemove.push(binding.path);
+                  } else {
+                    allElided = false;
+                    NEEDS_EXPLICIT_ESM.set(path.node, false);
+                  }
+                }
+
+                if (allElided) {
+                  stmt.remove();
+                } else {
+                  for (const importPath of importsToRemove) {
+                    importPath.remove();
+                  }
                 }
               }
-            }
 
-            continue;
-          }
+              continue;
+            }
 
-          if (stmt.isExportDeclaration()) {
-            stmt = stmt.get("declaration");
-          }
+            if (stmt.isExportDeclaration()) {
+              stmt = stmt.get("declaration");
+            }
 
-          if (stmt.isVariableDeclaration({ declare: true })) {
-            for (const name of Object.keys(stmt.getBindingIdentifiers())) {
-              registerGlobalType(path.scope, name);
+            if (stmt.isVariableDeclaration({ declare: true })) {
+              for (const name of Object.keys(stmt.getBindingIdentifiers())) {
+                registerGlobalType(path.scope, name);
+              }
+            } else if (
+              stmt.isTSTypeAliasDeclaration() ||
+              stmt.isTSDeclareFunction() ||
+              stmt.isTSInterfaceDeclaration() ||
+              stmt.isClassDeclaration({ declare: true }) ||
+              stmt.isTSEnumDeclaration({ declare: true }) ||
+              (stmt.isTSModuleDeclaration({ declare: true }) &&
+                stmt.get("id").isIdentifier())
+            ) {
+              registerGlobalType(path.scope, stmt.node.id.name);
             }
-          } else if (
-            stmt.isTSTypeAliasDeclaration() ||
-            stmt.isTSDeclareFunction() ||
-            stmt.isTSInterfaceDeclaration() ||
-            stmt.isClassDeclaration({ declare: true }) ||
-            stmt.isTSEnumDeclaration({ declare: true }) ||
-            (stmt.isTSModuleDeclaration({ declare: true }) &&
-              stmt.get("id").isIdentifier())
-          ) {
-            registerGlobalType(path.scope, stmt.node.id.name);
           }
-        }
+        },
+        exit(path) {
+          if (NEEDS_EXPLICIT_ESM.get(path.node)) {
```

Comment:
The tests are failing because here we also need to check if `sourceType` is still `module`: if it has been compiled to CJS we must not inject `export {}`.